Assignment 4/src/FractionalWrapper.sol

```diff
@@ -0,0 +1,318 @@
+    function mint(uint256 shares, address receiver) external override returns (uint256 assets) {
+        assets = convertToAssets(shares);
+        balanceOf[receiver] += shares;
+        totalSupply += shares;
+        if (token.transferFrom(msg.sender, address(this), assets)) {
+            emit Deposit(msg.sender, receiver, assets, shares);
+        } else {
+            revert DepositFailed();
+        }
+    }
+
+    /// @notice The maximal amount that can be withdrawn.
+    /// @param owner The withdrawer-to-be.
+    /// @return maxAssets The maximum amount of assets.
+    /// @dev The maximum is determined solely by the balance of the user.
+    function maxWithdraw(address owner) external override view returns (uint256 maxAssets) {
+        return convertToAssets(balanceOf[owner]);
+    }
+
+    /// @notice Preview how many shares would be minted with the given assets.
+    ///     For this contract, this function does the same as
+    ///     `convertToShares`.
+    /// @param assets How many assets would be put into the wrapper.
+    /// @return shares The shares that would be minted.
+    function previewWithdraw(uint256 assets) external override view returns (uint256 shares) {
+        return convertToShares(assets);
+    }
+
+    /// @notice Internal withdraw function. Used by `withdraw` and `redeem`
+    ///     since they are identical except for the computation of
+    ///     assets/shares.
+    /// @param assets The amount of assets to withdraw.
+    /// @param shares Shares corresponding to the assets.
+    /// @param receiver The receiver of the underlying.
+    /// @param owner The owner of the wrapper tokens.
+    /// @dev This function checks the following:
+    ///     - The caller should be the owner or have sufficient allowance.
+    ///         The allowance should be subtracted.
+    ///     - The balance of the owner in shares should be sufficient.
+    ///     - The token transfer should be successful.
+    ///     In all other cases the withdrawal should revert.
+    function _withdraw(uint256 assets, uint256 shares, address receiver, address owner) internal {
+        // First check ownership/permission:
+        if (msg.sender != owner) {
+            // Not the owner, try to subtract allowance.
+            uint256 _allowance = allowance[owner][msg.sender];
+            if (_allowance >= shares) {
+                allowance[owner][msg.sender] = _allowance - shares;
+            } else {
+                revert WithdrawalFailed();
+            }
+        }
+
+        // At this point the sender is the owner, or the allowance has been
+        // withdrawn.
+
+        // Check the balance
+        uint256 _balance = balanceOf[owner];
+        if (_balance < shares) {
+            revert WithdrawalFailed();
+        }
+
+        // Update the state
+        balanceOf[owner] = _balance - shares;
+        totalSupply -= shares;
+
+        if (token.transfer(receiver, assets)) {
```

Comment:
Not all ERC20 return false on failed transfers, despite being required in the EIP. Reverting on failed transfers is better UX and got widely adopted, to the point that some important tokens forgot about the return value. [Read this](https://decentraland.org/blog/technology/safe-erc20-transfers/).

Check [TransferHelper.sol](https://github.com/yieldprotocol/yield-utils-v2/blob/main/contracts/token/TransferHelper.sol), link it to IERC20 variables using `using`, and use the `.safeTransfer` function instead.